engine/src/main/java/org/terasology/entitySystem/event/AbstractBeforeAfterEvent.java

```diff
@@ -0,0 +1,42 @@
+// Copyright 2020 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.terasology.entitySystem.event;
+
+/**
+ * Immutable event to notify a change in a certain value.
+ * NOTE: This is a generic event and a more specific Event extending this event should be fired and reacted to.
+ */
+public class AbstractBeforeAfterEvent implements Event {
```

Comment:
Did you ever work with [Generics](https://www.baeldung.com/java-generics), as in, implemented a generic class? I'd think this would be a good case where we can use that - don't fix the event to be based on float right away.

So here's my suggestion:
```suggestion
public abstract class BeforeAfterEvent<T> implements Event {
```

- make the class abstract to ensure nobody can create a `new BeforeAfterEvent("foo", "bar")` out of thin air, but has to extend it to use it.
- shorten the name to `BeforeAfterEvent` - I don't see any good value in explicitly prefixing the "Abstract" part, and the inheritance line will read a little bit cleaner
  ```java
  public class HealthChangedEvent extends BeforeAfterEvent<Float> { ... }
  ```
- make the type of the changed value generic `<T>`. this allows to reuse the same base class regardless whether the value is an integer, a float, a string, or anything else. However, this implies that we cannot offer the `getChange` on this base class.

Reply:
Nope I certainly haven't worked with Generics 😅... And I did absolutely everything wrong I see 😅
And Now the component based alteration effect pr thing  (Class<C> componentClass) makes sense..
Thanks for explaining all this in such great detail!

Reply:
> And I did absolutely everything wrong

That's not true - basically everything was there, and it made it easy for me to add all those comments about fancy features :sweat_smile: I hope you'll learn one thing or the other, and please don't hesitate if you have questions (and please ask about things like `Class componentClass` if they are unclear :wink: )